packages/babel-generator/src/generators/typescript.ts

```diff
@@ -15,6 +15,7 @@ export function TSTypeParameterInstantiation(
 ): void {
   this.token("<");
   this.printList(node.params, node, {});
+  if(node.extra?.trailingComma) this.token(",");
```

Comment:
Thanks for that PR!
I think it would be safer to always print the comma if `node.params.length === 1`, so that it also works when the node has been modified/generated by a plugin. For example:
https://astexplorer.net/#/gist/ef9b1bdad45421b25b826a06f3a707e1/latest

Reply:
If I do it like that, now generator starts adding comma end of every conditional `node.params.length === 1` type inside every angle bracket. But some of those usages are not compatible with typescript, such as this one.

```javascript
// IN
// *****************************************************

const useState = <T>(x: T): T => x;

const foo = <T>(a: T) => {
  const x = useState<number>(2);
};


// OUT
// *****************************************************

const useState = <T,>(x: T): T => x;

const foo = <T,>(a: T) => {
  const x = useState<number,>(2); // now this becomes incompatible due to <number,>
};
```

Or am I missing something?

Reply:
Oh true. `TSTypeParameterInstantiation` is called with an additional argument, `parent: t.Node`: you can check if `parent.type === "ArrowFunctionExpression"`.

Reply:
We also need to find the type of  `FunctionDeclaration` right?

```typescript
function foo<T>(a: T) {} // AST says this is ArrowFunctionExpression (init type)
const foo= <T>(x: T): T => x; // and says this is FunctionDeclaration (type)
```

They both should be generated with at least one comma inside their angle brackets.

Reply:
I don't think the first one needs a trailing comma, because it's not ambiguous with JSX.

Reply:
> Oh true. `TSTypeParameterInstantiation` is called with an additional argument, `parent: t.Node`: you can check if `parent.type === "ArrowFunctionExpression"`.

I looked at many files but didn't find the 3th argument for `TSTypeParameterInstantiation` function, but then I found a way to check for if `node.type === TSTypeParameterDeclaration`. It seems reasonable for me to check it like so.

If this is not the proper solution, can you give me a hint that how can I reach the parent node inside `TSTypeParameterInstantiation` function?

Reply:
> Oh true. `TSTypeParameterInstantiation` is called with an additional argument, `parent: t.Node`: you can check if `parent.type === "ArrowFunctionExpression"`.

I looked at many files but didn't find the 3th argument for `TSTypeParameterInstantiation` function, but then I found a way to check for if `node.type === TSTypeParameterDeclaration`. It seems reasonable for me to check it like so.

If this is not the proper solution, can you give me a hint that how can I reach the parent node inside `TSTypeParameterInstantiation` function?

Reply:
Uh, doesn't just adding a third parameter to the function work?
```js
export function TSTypeParameterInstantiation(
  this: Printer,
  node: t.TSTypeParameterInstantiation | t.TSTypeParameterDeclaration,
  parent: t.Node
): void {
```
it's already called with two parameters (note that it's two and not three, because the first `this: Printer` is a "fake" TS parameter to annotate the `this` value).

Reply:
Example: https://github.com/ozanhonamlioglu/babel/blob/41b64db4a61a2f016ca44daade8192fc3b0ae467/packages/babel-generator/src/generators/flow.ts#L34-L38

Reply:
Oh gotcha, sending the update.

Reply:
Updated the PR, yes adding the third param does the job.
Can you give me one example, how `TSTypeParameterInstantiation` and other functions are called in generators folder, where the parameters are passed to those functions. Are there something like dependency injection?
Short answer is enough by the way, then I can search everything.

Reply:
Sure! You probably didn't find it because it's called indirectly, at https://github.com/babel/babel/blob/a6d77d07b461064deda6bdae308a0c70cacdd280/packages/babel-generator/src/printer.ts#L410 + https://github.com/babel/babel/blob/a6d77d07b461064deda6bdae308a0c70cacdd280/packages/babel-generator/src/printer.ts#L440

In this case `node.type` is `TSTypeParameterDeclaration`, so it calls the `TSTypeParameterDeclaration` export of the file you edited. And all these "printers" are installed as methods of the `Printer` class at https://github.com/babel/babel/blob/a6d77d07b461064deda6bdae308a0c70cacdd280/packages/babel-generator/src/printer.ts#L712. (I don't know why this design was chosen originally)

Reply:
Oh, I noticed the line 712 but then focused other parts. Thank you, I actually learned a lot in this work and absolutely will continue to work in the library. Thanks again.

Reply:
If you are interested, I have another bug for you which can be fixed after that this PR is merged (we first need a second review)! 

We currently parse `<T>() => {}` without problems even when both the JSX and TypeScript plugins are enabled. If there is a single type parameter and there isn't a trailing comma, we should report a parser error. It can probably be reported in https://github.com/babel/babel/blob/a6d77d07b461064deda6bdae308a0c70cacdd280/packages/babel-parser/src/plugins/typescript/index.js#L2861, using `this.raise` (see how it's used in other parts of that file). `OVERWRITE=true yarn jest parser` will then probably update some existing (wrong) tests: I would be surprised if we need new tests for this.

Reply:
Absolutely, I want to work on this case. please assign it to me.
When do you think other reviewer can review this PR?

Reply:
Probably @JLHwung or @existentialism can review this tomorrow.

I haven't created an issue for that bug; feel free to consider it assigned to you and just link my comment in the new PR :slightly_smiling_face: 

Btw, you don't need to wait for this to be merged because it's a parser bug, but you will then need to rebase your new PR after that this is merged to check if any related `@babel/generator` tests need to be merged.

Reply:
Absolutely, will start to work on it ASAP. Thanks @nicolo-ribaudo

Reply:
> If you are interested, I have another bug for you which can be fixed after that this PR is merged (we first need a second review)!
> 
> We currently parse `<T>() => {}` without problems even when both the JSX and TypeScript plugins are enabled. If there is a single type parameter and there isn't a trailing comma, we should report a parser error. It can probably be reported in
> 
> https://github.com/babel/babel/blob/a6d77d07b461064deda6bdae308a0c70cacdd280/packages/babel-parser/src/plugins/typescript/index.js#L2861
> 
> , using `this.raise` (see how it's used in other parts of that file). `OVERWRITE=true yarn jest parser` will then probably update some existing (wrong) tests: I would be surprised if we need new tests for this.
> (Note: that function is probably one of the most complex parser functions 😬)

Hi @nicolo-ribaudo 
I start to work on this bug today, will quote reply once I create a PR